src/main/java/io/thinkit/edc/client/connector/services/Service.java

```diff
@@ -0,0 +1,186 @@
+import java.io.InputStream;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.UnaryOperator;
+
+public class Service {
+    private final String url;
+    private final HttpClient httpClient;
+    private final UnaryOperator<HttpRequest.Builder> interceptor;
+
+    public Service(String url, HttpClient httpClient, UnaryOperator<HttpRequest.Builder> interceptor) {
+        this.url = url;
+        this.httpClient = httpClient;
+        this.interceptor = interceptor;
+    }
+
+    public String getUrl() {
+        return url;
+    }
+
+    public HttpClient getHttpClient() {
+        return httpClient;
+    }
+
+    public UnaryOperator<HttpRequest.Builder> getInterceptor() {
+        return interceptor;
+    }
+
+    protected <T, B> Result<T> send(
+            HttpRequest.Builder requestBuilder, B builder, BiFunction<HttpResponse<InputStream>, B, Result<T>> myFunc) {
+
+        try {
+            var request = interceptor.apply(requestBuilder).build();
+
+            var response = httpClient.send(request, HttpResponse.BodyHandlers.ofInputStream());
+            return myFunc.apply(response, builder);
+        } catch (IOException | InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    protected <T> Result<T> send(
+            HttpRequest.Builder requestBuilder, Function<HttpResponse<InputStream>, Result<T>> myFunc) {
+
+        try {
+            var request = interceptor.apply(requestBuilder).build();
+
+            var response = httpClient.send(request, HttpResponse.BodyHandlers.ofInputStream());
+            return myFunc.apply(response);
+        } catch (IOException | InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    protected <T, B> CompletableFuture<Result<T>> sendAsync(
+            HttpRequest.Builder requestBuilder, B builder, BiFunction<HttpResponse<InputStream>, B, Result<T>> myFunc) {
+
+        var request = interceptor.apply(requestBuilder).build();
+
+        CompletableFuture<HttpResponse<InputStream>> future =
+                httpClient.sendAsync(request, HttpResponse.BodyHandlers.ofInputStream());
+        return future.thenApply(response -> myFunc.apply(response, builder));
+    }
+
+    protected <T> CompletableFuture<Result<T>> sendAsync(
+            HttpRequest.Builder requestBuilder, Function<HttpResponse<InputStream>, Result<T>> myFunc) {
+
+        var request = interceptor.apply(requestBuilder).build();
+
+        CompletableFuture<HttpResponse<InputStream>> future =
+                httpClient.sendAsync(request, HttpResponse.BodyHandlers.ofInputStream());
+        return future.thenApply(myFunc);
+    }
+
+    protected <T> Result<T> getResponse(HttpResponse<InputStream> response, Function<JsonObject, T> build) {
```

Comment:
I see a common pattern in these `getXXX` methods: the way the response is handled is always the same: 
- check status code
  - if status code is successful extract body
  - else build error result (always in the same way)

my suggestion would be to extract a function that does exactly this. I think going down this path could make the signatures of `send` and `sendAsync` simpler